packages/expo-google-sign-in/android/src/main/java/expo/modules/google/signin/GoogleSignInModule.java

```diff
@@ -0,0 +1,449 @@
+            int code = e.getStatusCode();
+            authTask.reject(String.valueOf(code), GoogleSignInStatusCodes.getStatusCodeString(code));
+        }
+    }
+
+    @ExpoMethod
+    public void signInAsync(Promise promise) {
+        final GoogleSignInClient client = getClientOrReject(promise);
+        if (client == null) {
+            return;
+        }
+
+        if (authTask.update(promise, "signInAsync")) {
+            Boolean success = addUIBlockOrReject(new Runnable() {
+                @Override
+                public void run() {
+                    Intent signInIntent = client.getSignInIntent();
+
+                    Activity activity = getCurrentActivity();
+                    if (activity == null) {
+                        authTask.reject(ERROR_EXCEPTION, "activity is null");
+                        return;
+                    }
+                    activity.startActivityForResult(signInIntent, RC_LOG_IN);
+                }
+            }, promise);
+
+            if (!success) {
+                return;
+            }
+        }
+    }
+
+    @ExpoMethod
+    public void clearCacheAsync(final String token, Promise promise) {
+        try {
+            GoogleAuthUtil.clearToken(getApplicationContext(), token);
+            promise.resolve(null);
+        } catch (GoogleAuthException | IOException e) {
+            promise.reject(ERROR_EXCEPTION, e.getMessage(), e);
+        }
+    }
+
+    @ExpoMethod
+    public void signOutAsync(final Promise promise) {
+        GoogleSignInClient client = getClientOrReject(promise);
+
+        if (client == null) {
+            return;
+        }
+
+        client.signOut().addOnCompleteListener(new OnCompleteListener<Void>() {
+            @Override
+            public void onComplete(@NonNull Task<Void> task) {
+                handleSignOutOrRevokeAccessTask(task, promise);
+            }
+        });
+    }
+
+    @ExpoMethod
+    public void disconnectAsync(final Promise promise) {
+        GoogleSignInClient client = getClientOrReject(promise);
+        if (client == null) {
+            return;
+        }
+
+        client.revokeAccess().addOnCompleteListener(new OnCompleteListener<Void>() {
+            @Override
+            public void onComplete(@NonNull Task<Void> task) {
+                handleSignOutOrRevokeAccessTask(task, promise);
+            }
+        });
+    }
+
+    @ExpoMethod
+    public void isSignedInAsync(Promise promise) {
+        promise.resolve(GoogleSignIn.getLastSignedInAccount(getApplicationContext()) != null);
+    }
+
+    private void handleSignOutOrRevokeAccessTask(@NonNull Task<Void> task, final Promise promise) {
+        if (task.isSuccessful()) {
+            promise.resolve(true);
+        } else {
+            int code = getExceptionCode(task);
+            promise.reject(String.valueOf(code), GoogleSignInStatusCodes.getStatusCodeString(code));
+        }
+    }
+
+    private boolean addUIBlockOrReject(Runnable block, Promise promise) {
```

Comment:
maybe consider calling this `addUIThreadRunnableOrReject` -- `block` is more of an obj c term